src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/TypeBuilderImpl.cs

```diff
@@ -230,7 +231,110 @@ protected override MethodBuilder DefineMethodCore(string name, MethodAttributes
             return methodBuilder;
         }
 
-        protected override void DefineMethodOverrideCore(MethodInfo methodInfoBody, MethodInfo methodInfoDeclaration) => throw new NotImplementedException();
+        protected override void DefineMethodOverrideCore(MethodInfo methodInfoBody, MethodInfo methodInfoDeclaration)
+        {
+            ThrowIfCreated();
+            if (!ReferenceEquals(methodInfoBody.DeclaringType, this))
+            {
+                // Loader restriction: body method has to be from this class
+                throw new ArgumentException(SR.ArgumentException_BadMethodImplBody);
+            }
+            Type baseType = methodInfoDeclaration.DeclaringType!;
+            ValidateBaseType(baseType, methodInfoDeclaration.Name);
+            ValidateImplementedMethod(methodInfoBody, methodInfoDeclaration);
+
+            _interfaceMappings ??= new();
+            if (_interfaceMappings.TryGetValue(baseType, out InterfaceMap? im))
+            {
+                if (im._interfaceMethods.Contains(methodInfoDeclaration))
+                {
+                    throw new ArgumentException(SR.Format(SR.Argument_MethodOverriden, methodInfoBody.Name, FullName), nameof(methodInfoDeclaration));
+                }
+
+                im._interfaceMethods.Add(methodInfoDeclaration);
+                im._targetMethods.Add(methodInfoBody);
+            }
+            else
+            {
+                im = new InterfaceMap(baseType);
+                im._interfaceMethods.Add(methodInfoDeclaration);
+                im._targetMethods.Add(methodInfoBody);
+                _interfaceMappings.Add(baseType, im);
+            }
+        }
+
+        private void ValidateBaseType(Type baseType, string methodName)
+        {
+            if (baseType.IsInterface)
+            {
+                if (!IsInterfaceImplemented(baseType))
+                {
+                    throw ArgumentExceptionInvalidMethodOverride(methodName);
+                }
+
+                return;
+            }
+
+            if (!baseType.IsAssignableFrom(_typeParent))
+            {
+                throw ArgumentExceptionInvalidMethodOverride(methodName);
+            }
+        }
+
+        private bool IsInterfaceImplemented(Type interfaceType)
+        {
+            if (_interfaces != null)
+            {
+                foreach (Type @interface in _interfaces)
```

Comment:
I thought using keywords in the BCL was frowned upon? There are more than enough ways to express interface without using that word. Perhaps `iface`?